src/filter.rs

```diff
@@ -42,29 +55,71 @@ impl<'a, 'b> RayTracing<'a, 'b> {
     /// in `[-1, 1]`.
     pub fn albedo_normal(
         &mut self,
-        albedo: &'b [f32],
-        normal: &'b [f32],
-    ) -> &mut RayTracing<'a, 'b> {
-        self.albedo = Some(albedo);
-        self.normal = Some(normal);
+        albedo: &[f32],
+        normal: &[f32],
+    ) -> &mut RayTracing<'a> {
+        match self.albedo {
+            None => {
+                unsafe {
+                    oidnRetainDevice(self.device.0);
+                }
+                let byte_size = albedo.len() * 4;
+                let albedo_buffer = unsafe { oidnNewBuffer(self.device.0, byte_size) };
+                unsafe { oidnWriteBuffer(albedo_buffer, 0, byte_size, albedo.as_ptr() as *const _) }
+                self.albedo = Some((albedo_buffer, albedo.len()));
+            }
+            Some((buffer, _)) => {
+                unsafe { oidnWriteBuffer(buffer, 0, albedo.len() * 4, albedo.as_ptr() as *const _) }
+                self.albedo = Some((buffer, albedo.len()));
+            }
+        }
+        match self.normal {
+            None => {
+                unsafe {
+                    oidnRetainDevice(self.device.0);
+                }
+                let byte_size = normal.len() * 4;
+                let normal_buffer = unsafe { oidnNewBuffer(self.device.0, byte_size) };
+                unsafe { oidnWriteBuffer(normal_buffer, 0, byte_size, normal.as_ptr() as *const _) }
+                self.normal = Some((normal_buffer, normal.len()));
+            }
+            Some((buffer, _)) => {
+                unsafe { oidnWriteBuffer(buffer, 0, normal.len() * 4, normal.as_ptr() as *const _) }
+                self.normal = Some((buffer, normal.len()));
+            }
+        }
         self
     }
 
     /// Set an input auxiliary image containing the albedo per pixel (three
     /// channels, values in `[0, 1]`).
-    pub fn albedo(&mut self, albedo: &'b [f32]) -> &mut RayTracing<'a, 'b> {
-        self.albedo = Some(albedo);
+    pub fn albedo(&mut self, albedo: &[f32]) -> &mut RayTracing<'a> {
+        match self.albedo {
+            None => {
+                unsafe {
+                    oidnRetainDevice(self.device.0);
```

Comment:
I don't think these `oidnRetainDevice` calls are needed when making buffers (here and in `albedo_normal` since the `RayTracing` filter already owns a reference to the device, these additional retains would be redundant. Could we try removing them?

Reply:
Thanks for your feedback! Rechecking Intel documentation, it would appear that I have misunderstood what that function did. I have fixed it now and updated my pull request.